compiler/rustc_middle/src/ty/consts.rs

```diff
@@ -398,133 +398,65 @@ impl<'tcx> Const<'tcx> {
         }
     }
 
-    /// Returns the evaluated constant
-    #[inline]
-    pub fn eval(
-        self,
-        tcx: TyCtxt<'tcx>,
-        param_env: ParamEnv<'tcx>,
-        span: Span,
-    ) -> Result<(Ty<'tcx>, ValTree<'tcx>), ErrorHandled> {
-        self.eval_valtree(tcx, param_env, span).map_err(|err| {
-            match err {
-                Either::Right(err) => err,
-                Either::Left(_bad_ty) => {
-                    // This can happen when we run on ill-typed code.
-                    let e = tcx.dcx().span_delayed_bug(
-                        span,
-                        "`ty::Const::eval` called on a non-valtree-compatible type",
-                    );
-                    e.into()
-                }
-            }
-        })
-    }
-
     /// Normalizes the constant to a value or an error if possible.
```

Comment:
This comment should explain what makes the function "internal", and what callers should use instead.

Reply:
Sure, I can leave a brief comment but essentially this function should only be called within the type-system normalization functions themselves. Unfortunately, with the new solver there are currently three normalization routines (AssocNormalizer, QueryNormalizer, and the new solver) so it was easier to deduplicate them here.

I can lift that into the comment, and advise users to use `normalize_erasing_regions` if necessary, but that it should also be unnecessary to call anything like `normalize` at all if the const is coming from mir which has been passed through `instantiate_mir_and_normalize_erasing_regions`, similarly how mir-opts and codegen aren't sprinkled with tons of calls to `normalize_erasing_regions` for `Ty`.

The only time that you *need* to normalize something is when it's coming directly from the type system queries like `type_of`, etc., but that's already *hopefully* being normalized because codegen already doesn't know what the heck to do with unnormalized *types* :3

Reply:
Ah okay I think I finally get it. The goal is to generally make it an invariant that constants are in their normalized (i.e., evaluated) form, like it is already the case for types.

Does the existing normalization for `mir::Const` already do that?

Reply:
That would mean we should probably rename `eval_mir_constant` in the interpreter (and codegen?) to something that does not involve "eval". All it does then is check the constant it got, and any `Unevaluated` case immediately leads to `TooGeneric`. We have `const_val_to_op`, I guess this would be `const_to_op`?

Reply:
> Does the existing normalization for mir::Const already do that?

Well, `mir::Const::normalize` works as it should, AFAICT. It just, unfortunately, is kinda necessary to still call since normalization of `mir::Const` **doesn't** happen implicitly during substitution-and-instantation that monomorphization does in CTFE and codegen, unlike `ty::Const` (especially after this work is done).

Reply:
`mir::Const::normalize` isn't really necessary though... Like, who cares about *trying* to normalize a mir const? We're not putting it back into anything. All callsites should be calling `mir::Const::eval`, probably. 

`mir::Const::normalize` or something morally equivalent would probably be used if we made the `NormalizeErasingRegions` folder actually fold MIR consts like it folds ty consts, though.

Reply:
>  normalization of mir::Const doesn't happen implicitly during substitution-and-instantation

That's what I meant. Should this happen? Seems a bit odd not to do it, but there may well be good reasons.

I assume the `ty::Const` inside a `mir::Const` *does* get fully evaluated, so this is only about `mir::Const::Unevaluated`?

> All callsites should be calling mir::Const::eval, probably.

It's possible that this is just a holdover from when nothing like `eval` existed, and in introducing `eval` I didn't want to port all the existing normalization-based code.

Reply:
> I assume a ty::Const inside a mir::Const does get fully evaluated, so this is only about mir::Const::Unevaluated?

Correct

> It's possible that this is just a holdover from when nothing like eval existed, and in introducing eval I didn't want to port all the existing normalization-based code.

Perhaps. I actually looked, and the only place I could find that uses normalization is a call that can be replaced wiht `try_eval_target_usize` in jump threading, and some `normalize` calls I added to `mir_build` that *try* to eagerly eval a `const` argument to `asm!()` macro to eagerly raise errors in polymorphic functions like:

```
fn not_instantiated<T>() {
    unsafe {
        asm!("/* {} */", const 1 / 0);
        //~^ ERROR evaluation of const failed
    }
}
```

Which is actually inconsistent with things like inline consts:

```
fn test<T>() {
    let x = const { 1 / 0 };
}
```

which doesn't raise an error until it's monomorphized 😆

Reply:
For the record, `normalize` on `mir::Const` is gone now: https://github.com/rust-lang/rust/pull/130990